src/lib/async-append.ts

```diff
@@ -0,0 +1,88 @@
+/**
+ * @license
+ * Copyright (c) 2017 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {directive, NodePart} from '../lit-html.js';
+
+/**
+ * A directive that renders the items of an async iterable[1], appending new
+ * values after previous values, similar to the built-in support for iterables.
+ *
+ * Async iterables are objects with a [Symbol.asyncIterator] method, which
+ * returns an iterator who's `next()` method returns a Promise. When a new
+ * value is available, the Promise resolves and the value is appended to the
+ * Part controlled by the directive. If another value other than this
+ * directive has been set on the Part, the iterable will no longer be listened
+ * to and new values won't be written to the Part.
+ *
+ * [1]: https://github.com/tc39/proposal-async-iteration
+ *
+ * @param value An async iterable
+ * @param mapper An optional function that maps from (value, index) to another
+ *     value. Useful for generating templates for each item in the iterable.
+ */
+export const asyncAppend = <T>(
+    value: AsyncIterable<T>, mapper?: (v: T, index?: number) => any) =>
+    directive(async (part: NodePart) => {
+      // If we've already set up this particular iterable, we don't need
+      // to do anything.
+      if (value === part._previousValue) {
+        return;
+      }
+      part.clear();
```

Comment:
Can we wait to clear? That way we could use something like `until` to get a default value while the first promise is resolving.

Reply:
Good point, and I'll do that, but...

I think a lot of the directive, especially when dealing with async values, don't compose well right now. `until(asyncIterable, fallback)` won't quite work. I could spot-fix a few things, but I want to think on this a little and see if there's a more general solution.